### DISCUSSION TITLE
pipeline integration test for pipeline with parameters

### DISCUSSION THREAD

**Discussion by martinlarsalbert** [kedro-org/kedro#1068] [Idea]
How should I write an integration test for my pipelines?

I have managed to write an integration test here:
https://gist.github.com/martinlarsalbert/dd1aef5d63902c55e24f54c1a4523018#file-test_pipeline-py

I'm not sure if this is the correct way? (But it works anyway)

My problem now is that my pipeline also has parameters (from the parameters.yml file) how should I incorporate that into the integration test?

**Comment by datajoely** [comment]
Hi @martinlarsalbert this was recently discussed on our [Discord](https://discord.gg/akJDeVaxnB) server in [this thread](https://discord.com/channels/778216384475693066/778998585454755870/864551888137486336). 

I'll copy the answer here for visibility:

> A few guidelines I try to stick to:
> 
> 1. Unit test everything that is not inside pipeline.py. 
>    * Treat these like you would any other unit test you do. 
>    * Use tiny, targeted data per-test to keep your test suite fast. Usually single-row datasets work fine for unit tests, but occasionally you may stretch it to 10 randomly generated ones.
> 2. Integration test everything that is inside pipeline.py
>    * This is where you typically stitch functions together. You already know they work individually from your unit tests, so completely disregard any form of unit testing for pipelines. Instead, think of them as an integration.
>    * Provide free inputs
>    * Test on final outputs
>    * Completely ignore intermediates - you already unit tested them.
> 3. System test the entire pipeline.
>    * Also known as "end-to-end" testing, or "automated acceptance" testing.
>    * Your test runner will losely look like:
>       (cd source_proj & kedro pipeline package my_pipeline)
>       (cd test_dir & kedro new --config /path/to/test/config)
>       (cd test_dir/test_proj & kedro pipeline pull source_proj/dist/my_pipeline*.whl)
>       ... any other data / config / pipelines that need set up ...
>       kedro run
> 4. Never ever use catalog.yml or parameters.yml or data/ files in your tests.
>    * For unit tests, you'll likely want to try out many variations of parameters, so you can't anyway!
>    * Use kedro code API instead
> 5. Make use of pytest features, they make life a lot easier
>    * use fixtures for setting up default catalogs and parameters (top-tip: pytest has a built-in fixture tmp_path - use that in your catalog entries)
>    * conftest.py is a really useful file

**Comment by idrisrupt** [comment]
Hi @martinlarsalbert like @datajoely said it's not recommended to use parameters.yml
instead i used another dictionnary in the datacatalog initialised with code api.
So you have to add your parameters to the feed_dict dictionnary.
Good Luck.
![image](https://user-images.githubusercontent.com/30355857/150945875-5b2d8c84-a43c-4bb9-9170-e98d610c05f5.png)

**Reply by datajoely** [reply]
Since this is active again - here is a little demo I saw recently, currently thinking how to build it into the template:

![image](https://user-images.githubusercontent.com/35801847/150944511-6dbd96f3-6788-40ee-8ef2-f918810ea425.png)

**Reply by selalamiTF** [reply]
Hello, Thanks for sharing ! as this is very interesting, would you be able to share the whole demo ? Thanks !

**Reply by datajoely** [reply]
I can't today - but keen to update the PR some ideas into the default template when I get a chance

**Reply by allarddewinter** [reply]
@datajoely: I am not sure if I understand the example correctly: Comparing this example and the [previous reply you wrote above](https://github.com/kedro-org/kedro/discussions/1068#discussioncomment-1699371), being inside a pipeline this example is clearly about tests in step 2. But then if you would have tested everything outside pipeline.py, for example functions in nodes.py, then you would not have to test every single step in the pipeline, or am I missing something? 

I would rather expect that in this example you would test the pipeline as a whole in a single test, from the first node to the last, so you know that this single pipeline (all nodes together) also function as expected.

And the difference between step 2 and 3 would be that in step 3 you would test several pipelines together that are combined in the code you created (like in the pipeline_registry).

**Reply by datajoely** [reply]
I this example you are importing the nodes from `nodes.py`. 

To test individual node units, you should write individual unit tests. To test how they integrate together, you should write integration tests to confirm they flow correctly.

**Comment by selalamiTF** [comment]
Thanks a lot ! Will be following the update

Téléchargez Outlook pour iOS<https://aka.ms/o0ukef>
________________________________
De : Joel Schwarzmann ***@***.***>
Envoyé : Tuesday, February 8, 2022 4:51:49 PM
À : kedro-org/kedro ***@***.***>
Cc : Soufiane El Alami ***@***.***>; Comment ***@***.***>
Objet : Re: [kedro-org/kedro] pipeline integration test for pipeline with parameters (Discussion #1068)

Message from External Sender


I can't today - but keen to update the PR some ideas into the default template when I get a chance

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/kedro-org/kedro/discussions/1068*discussioncomment-2134352__;Iw!!A1Yki2KVLpWc!uy-5336HvGVxsncDEYD_eBcDmPCvgIvR9YwCaXu_C6uCjACXQS-J-pjiNu3L-5k$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/ASXTHC4UAICXLCXMS5C2J7TU2E3ZLANCNFSM5IYFTJRQ__;!!A1Yki2KVLpWc!uy-5336HvGVxsncDEYD_eBcDmPCvgIvR9YwCaXu_C6uCjACXQS-J-pjiH3galr0$>.
Triage notifications on the go with GitHub Mobile for iOS<https://urldefense.com/v3/__https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675__;!!A1Yki2KVLpWc!uy-5336HvGVxsncDEYD_eBcDmPCvgIvR9YwCaXu_C6uCjACXQS-J-pjixWOrEvA$> or Android<https://urldefense.com/v3/__https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign*3Dnotification-email*26utm_medium*3Demail*26utm_source*3Dgithub__;JSUlJSU!!A1Yki2KVLpWc!uy-5336HvGVxsncDEYD_eBcDmPCvgIvR9YwCaXu_C6uCjACXQS-J-pji5hm084I$>.

**Comment by JoseRZapata** [comment]
To load parameters in the test function you can get it with this code:

```python
import os
from kedro.config import ConfigLoader

loader = ConfigLoader(
            conf_source=os.path.abspath(path="conf"),
            env="base"
        )
conf_parameters = loader.get("parameters*",
                                     "parameters*/**",
                                     "**/parameters*")
```
